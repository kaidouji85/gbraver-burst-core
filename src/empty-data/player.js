// @flow

import type {PlayerState} from "../state/player-state";
import {EMPTY_ARMDOZER_STATE} from "./armdozer";
import {EMPTY_PILOT_STATE} from "./pilot";

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: '',
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
