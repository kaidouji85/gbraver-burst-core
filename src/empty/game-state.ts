import type { GameState } from "../state/game-state";

/** 空のゲームステート */
export const EMPTY_GAME_STATE: GameState = {
  players: [],
  activePlayerId: "",
  effect: {
    name: "StartGame"
  }
};