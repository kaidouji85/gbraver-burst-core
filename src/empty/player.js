// @flow

import type { Player } from "../player/player";
import type { PlayerState } from "../state/player-state";
import { EMPTY_ARMDOZER, EMPTY_ARMDOZER_STATE } from "./armdozer";
import { EMPTY_PILOT, EMPTY_PILOT_STATE } from "./pilot";

/** 空のプレイヤー */
export const EMPTY_PLAYER: Player = {
  playerId: "",
  armdozer: EMPTY_ARMDOZER,
  pilot: EMPTY_PILOT,
};

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: "",
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
