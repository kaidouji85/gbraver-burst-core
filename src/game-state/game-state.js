// @flow
import type {SecretGameState} from "./secret-game-state";
import type {OpenGameState} from "./open-game-state";

/** ゲームの状態 */
export type GameState = {
  /** 全プレイヤーに公開可能な状態 */
  openState: OpenGameState,
  /** 特定のプレイヤーにしか公開しない状態 */
  secretState: SecretGameState
};