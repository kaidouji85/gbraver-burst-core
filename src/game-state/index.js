import type {GameStateStep} from "./game-state-step";

/** 現在のゲーム状態 */
export type GameState = {
  steps: GameStateStep[]
};
