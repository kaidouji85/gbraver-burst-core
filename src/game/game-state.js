import type {PlayerGameState} from "../player/player-game-state";
import type {PlayerId} from "../player/player";
import {PhaseName} from "./phase-name";

/** 現在のゲーム状態 */
export type GameState = {
  /** プレイヤー毎の状態 */
  players: PlayerGameState[],
  /** 攻撃側のプレイヤーのID */
  activePlayerId: PlayerId,
  /** フェイズ名 */
  phase: PhaseName,
};