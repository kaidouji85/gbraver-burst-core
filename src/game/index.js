// @flow

import type {Player} from "../player/player";
import type {GameState} from "../state/game-state";
import type {PlayerCommand} from "../player/command/player-command";
import {start} from "./start/start";
import {progress} from "./progress";

/**
 * Gブレイバーバーストのコア
 *
 */
export class GbraverBurstCore {
  /**
   * ゲームの初期状態を生成する
   *
   * @param player1 プレイヤー1
   * @param player2 プレイヤー2
   * @return ゲーム初期状態
   */
  start(player1: Player, player2: Player): GameState[] {
    return start(player1, player2);
  }

  /**
   * ゲームを進める
   *
   * @param lastState 最後の状態
   * @param commands コマンド
   * @return 更新されたゲーム状態
   */
  progress(lastState: GameState, commands: PlayerCommand[]): GameState[] {
    return progress(lastState, commands);
  }
}