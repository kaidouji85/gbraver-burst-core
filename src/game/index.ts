import type { Player } from "../player/player";
import type { GameState } from "../state/game-state";
import type { PlayerCommand } from "./command/player-command";
import { progress } from "./progress";
import { RestoreGBraverBurst } from "./restore-gbraver-burst";
import { start } from "./start/start";
import { isAllPlayerEnteredCommand } from "./validation/is-all-player-entered-command";
import { isDuplicatePlayers } from "./validation/is-duplicate-players";
import { isValidCommand } from "./validation/is-valid-command";


/** Gブレイバーバーストコア */
export interface GBraverBurstCore {
  /**
   * バトルに参加している全プレイヤーを取得する
   * @return 取得結果
   */
  players(): [Player, Player];

  /**
   * ゲームステート履歴を取得する
   * @return 取得結果
   */
  stateHistory(): GameState[];

  /**
   * 現在の状態をダンプする
   * @return ダンプしたデータ
   */
  dump(): RestoreGBraverBurst;

  /**
   * ゲームを進行させる
   * @param commands コマンド
   * @return 更新されたゲーム状態
   */
  progress(commands: [PlayerCommand, PlayerCommand]): GameState[];
}

/** Gブレイバーバーストコア実装 */
class GBraverBurstCoreImpl implements GBraverBurstCore {
  #players: [Player, Player];
  #stateHistory: GameState[];

  /**
   * コンストラクタ
   *
   * @param players バトルに参加するプレイヤー
   * @param stateHistory ステートヒストリー
   */
  constructor(players: [Player, Player], stateHistory: GameState[]) {
    this.#players = players;
    this.#stateHistory = stateHistory;
  }

  /** @override */
  players(): [Player, Player] {
    return this.#players;
  }

  /** @override */
  stateHistory(): GameState[] {
    return this.#stateHistory;
  }

  /** @override */
  dump(): RestoreGBraverBurst {
    return {
      players: this.#players,
      stateHistory: this.#stateHistory,
    };
  }

  /** @override */
  progress(commands: [PlayerCommand, PlayerCommand]): GameState[] {
    if (!isAllPlayerEnteredCommand(this.#players, commands)) {
      throw new Error("all player not enter command");
    }

    const lastState = this.#stateHistory[this.#stateHistory.length - 1];
    if (!lastState) {
      throw new Error("no game state history");
    }

    const isValidCommands =
      lastState.effect.name === "InputCommand" &&
      isValidCommand(commands[0], lastState.effect) &&
      isValidCommand(commands[1], lastState.effect);
    if (!isValidCommands) {
      throw new Error("invalid commands");
    }

    const updated = progress(lastState, commands);
    this.#stateHistory = [...this.#stateHistory, ...updated];
    return updated;
  }
}

/**
 * Gブレイバーバーストを開始する
 * @param players プレイヤー情報
 * @return Gブレイバーバースト
 */
export function startGBraverBurst(players: [Player, Player]): GBraverBurstCore {
  if (isDuplicatePlayers(players)) {
    throw new Error("duplicate players");
  }

  const stateHistory = start(players);
  return new GBraverBurstCoreImpl(players, stateHistory);
}

/**
 * Gブレイバーバーストを再開する
 * @param data 再開するデータ
 * @return Gブレイバーバースト
 */
export function restoreGBraverBurst(
  data: RestoreGBraverBurst,
): GBraverBurstCore {
  return new GBraverBurstCoreImpl(data.players, data.stateHistory);
}