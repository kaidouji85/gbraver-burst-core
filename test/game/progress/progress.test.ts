import path from "path";
import type { BatteryCommand, BurstCommand, GameState, PlayerState } from "../../../src";
import type { PilotSkillCommand } from "../../../src/command/pilot-skill";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";
import { progress } from "../../../src/game/progress";
import { exportSnapShotJSON, importSnapShotJSON, shouldUpdateSnapShot } from "../../snap-shot";
const ATTACKER: PlayerState = { ...EMPTY_PLAYER_STATE,
  playerId: "attacker"
};
const DEFENDER: PlayerState = { ...EMPTY_PLAYER_STATE,
  playerId: "defender"
};
const BATTERY_COMMAND: BatteryCommand = {
  type: "BATTERY_COMMAND",
  battery: 1
};
const BURST_COMMAND: BurstCommand = {
  type: "BURST_COMMAND"
};
const PILOT_SKILL_COMMAND: PilotSkillCommand = {
  type: "PILOT_SKILL_COMMAND"
};
test("戦闘フローを正しく進めることができる", () => {
  const state: GameState = { ...EMPTY_GAME_STATE,
    players: [DEFENDER, ATTACKER],
    activePlayerId: ATTACKER.playerId
  };
  const commands = [{
    playerId: ATTACKER.playerId,
    command: BATTERY_COMMAND
  }, {
    playerId: DEFENDER.playerId,
    command: BATTERY_COMMAND
  }];
  const result = progress(state, commands);
  const snapShotPath = path.join(__dirname, "progress__battle-flow.json");
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot() ? result : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});
test("効果適用フローを正しく進めることができる", () => {
  const state: GameState = { ...EMPTY_GAME_STATE,
    players: [DEFENDER, ATTACKER],
    activePlayerId: ATTACKER.playerId
  };
  const commands = [{
    playerId: ATTACKER.playerId,
    command: BURST_COMMAND
  }, {
    playerId: DEFENDER.playerId,
    command: PILOT_SKILL_COMMAND
  }];
  const result = progress(state, commands);
  const snapShotPath = path.join(__dirname, "progress__effect-activation-flow.json");
  shouldUpdateSnapShot() && exportSnapShotJSON(snapShotPath, result);
  const snapShot = shouldUpdateSnapShot() ? result : importSnapShotJSON(snapShotPath);
  expect(result).toEqual(snapShot);
});