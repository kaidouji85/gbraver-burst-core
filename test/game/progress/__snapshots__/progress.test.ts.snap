// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`効果適用フローを正しく進めることができる: effect-activation-flow 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "burst": {
        "recoverBattery": 5,
        "turnStartBatteryCorrect": 1,
        "type": "RecoverBattery",
      },
      "burstPlayer": "attacker",
      "name": "BurstEffect",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "invokerId": "defender",
      "name": "PilotSkillEffect",
      "skill": {
        "recoverBattery": 2,
        "type": "RecoverBatterySkill",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": false,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "attacker",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
          ],
          "playerId": "defender",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": false,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
]
`;

exports[`戦闘フローを正しく進めることができる: battle-flow 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "attackerBattery": 1,
      "defenderBattery": 1,
      "name": "BatteryDeclaration",
      "originalBatteryOfAttacker": 1,
      "originalBatteryOfDefender": 1,
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "isDeath": false,
      "name": "Battle",
      "result": {
        "damage": 1000,
        "name": "Guard",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "battleResult": {
        "damage": 1000,
        "name": "Guard",
      },
      "defender": "defender",
      "name": "RightItself",
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "endArmdozerEffects": [],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "defender",
    "effect": {
      "name": "TurnChange",
      "recoverBattery": 3,
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "defender",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "defender",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "attacker",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
]
`;
